addons/mail/static/src/mail_message.js

```diff
@@ -0,0 +1,148 @@
+    // Adapted to make http(s):// not required if (and only if) www. is given. So `should.notmatch` does not match.
+    // And further extended to include Latin-1 Supplement, Latin Extended-A, Latin Extended-B and Latin Extended Additional.
+    const escapedAndCompactContent = escapeAndCompactTextContent(rawBody);
+    let body = escapedAndCompactContent.replace(/&nbsp;/g, " ").trim();
+    // This message will be received from the mail composer as html content
+    // subtype but the urls will not be linkified. If the mail composer
+    // takes the responsibility to linkify the urls we end up with double
+    // linkification a bit everywhere. Ideally we want to keep the content
+    // as text internally and only make html enrichment at display time but
+    // the current design makes this quite hard to do.
+    // body = this._generateMentionsLinks(body);
+    body = parseAndTransform(body, addLink);
+    body = _generateEmojisOnHtml(body);
+    return body;
+}
+
+/**
+ * WARNING: this is not enough to unescape potential XSS contained in htmlString, transformFunction
+ * should handle it or it should be handled after/before calling parseAndTransform. So if the result
+ * of this function is used in a t-raw, be very careful.
+ *
+ * @param {string} htmlString
+ * @param {function} transformFunction
+ * @returns {string}
+ */
+function parseAndTransform(htmlString, transformFunction) {
+    var openToken = "OPEN" + Date.now();
+    var string = htmlString.replace(/&lt;/g, openToken);
+    var children;
+    try {
+        children = $("<div>").html(string).contents();
+    } catch {
+        children = $("<div>")
+            .html("<pre>" + string + "</pre>")
+            .contents();
+    }
+    return _parseAndTransform(children, transformFunction).replace(
+        new RegExp(openToken, "g"),
+        "&lt;"
+    );
+}
+
+/**
+ * @param {Node[]} nodes
+ * @param {function} transformFunction with:
+ *   param node
+ *   param function
+ *   return string
+ * @return {string}
+ */
+function _parseAndTransform(nodes, transformFunction) {
+    return _.map(nodes, function (node) {
+        return transformFunction(node, function () {
+            return _parseAndTransform(node.childNodes, transformFunction);
+        });
+    }).join("");
+}
+
+/**
+ * @param {string} text
+ * @param {Object} [attrs={}]
+ * @return {string} linkified text
+ */
+function linkify(text, attrs) {
+    attrs = attrs || {};
+    if (attrs.target === undefined) {
+        attrs.target = "_blank";
+    }
+    if (attrs.target === "_blank") {
+        attrs.rel = "noreferrer noopener";
+    }
+    attrs = _.map(attrs, function (value, key) {
+        return key + '="' + _.escape(value) + '"';
+    }).join(" ");
+    return text.replace(urlRegexp, function (url) {
+        var href = !/^https?:\/\//i.test(url) ? "http://" + url : url;
+        return "<a " + attrs + ' href="' + href + '">' + url + "</a>";
```

Comment:
again no `_.map()` or `_.escape()` (for later, use imported `escape()` instead), or `var`